### DISCUSSION TITLE
Server to Server OAuth ZoomNet question.

### DISCUSSION THREAD

**Discussion by jwjhhcjc** [Jericho/ZoomNet#230] [Q&A]
Hello,
I need to change an win form application from JWT to the new Server to Server OAuth. I am using ZoomNet version v0.46.0 and have the following vb.net code 4.8 framework. I have set up a Server to Server application on Zoom and acquired the client id, Account id and Client secret and entered all the information into the variables. I get a token when testing using Postman with the information from Zoom. However, when I try the following code I get no errors but no token is returned. Any help is greatly appreciated!

```
Try
 	Dim sClientID As String = "redacted"
        Dim sClientSecret As String = "redacted"
        Dim sAccountID As String = "redacted"
        Dim RefreshToken As OnTokenRefreshedDelegate = Nothing
        Dim sAccessToken As String

        Dim ZoomConnInfo As New OAuthConnectionInfo(sClientID, sClientSecret, sAccountID, RefreshToken) 
        
            sAccessToken  = ZoomConnInfo.AccessToken
        
            ZoomConnInfo = Nothing
Catch ex As Exception
           MessageBox.Show(ex.Message)           
End Try
```
When I debug print ZoomConnInfo I see the following.
```
?ZoomConnInfo
{ZoomNet.OAuthConnectionInfo}
    AccessToken: Nothing
    AccountId: "redacted"
    AuthorizationCode: Nothing
    ClientId: "redacted"
    ClientSecret: "redacted"
    GrantType: AccountCredentials {0}
    OnTokenRefreshed: Nothing
    RedirectUri: Nothing
    RefreshToken: Nothing
    TokenExpiration: #1/1/0001 12:00:00 AM#
    TokenScope: Nothing
```

**Comment by Jericho** [comment]
The `OAuthConnectionInfo` class is not the class responsible for obtaining or refreshing a token. As its name implies, it merely contains "info" about the OAuth connection such as the clientId, secret, accountId, etc. The Zoom Client will, when appropriate, obtain a new token (or refresh an expired token)  and at that time it will update other properties of the connection info object such as expiration date, scopes, refresh token, etc.

Therefore, it's perfectly normal for properties such as the token expiration date, access token, refresh token, etc. to contain null values ("Nothing" in VB.NET) as you have shown in your code example.

**Comment by jwjhhcjc** [comment]
Thank you for responding so quickly! Thank you for the clarification. Is there a class that will accept my credentials and return a token in ZoomNet or is that done via some other way?

Thank you!

**Comment by Jericho** [comment]
The ZoomNet client is designed to take care of token management for you and also, it is designed to raise an event to notify you when a new token is obtained if that's something you are interested in. Here's an example that shows how you would specify the logic you want to execute whenever a token is obtained or refreshed:
```csharp
var connectionInfo = new OAuthConnectionInfo(clientId, clientSecret, refreshToken, accessToken,
    (newRefreshToken, newAccessToken) =>
    {
        Environment.SetEnvironmentVariable("ZOOM_OAUTH_REFRESHTOKEN", newRefreshToken, EnvironmentVariableTarget.User);
        Environment.SetEnvironmentVariable("ZOOM_OAUTH_ACCESSTOKEN", newAccessToken, EnvironmentVariableTarget.User);
    });
```
In this example, I am updating two environment variables to preserve the refresh token as well as the access token. Of course, this logic can be tailored to match your needs.

What this means is that you shouldn't need to manually obtain a new token, you simply let ZoomNet take care of it for you.

**Comment by jwjhhcjc** [comment]
Thank you for getting me on the right track!!
